src/Parser.php

```diff
@@ -363,18 +360,31 @@ protected function isIpv6Host(string $ipv6): bool
      */
     protected function isRegisteredName(string $host): bool
     {
-        if ('.' === substr($host, -1, 1)) {
-            $host = substr($host, 0, -1);
+        // Note that unreserved is purposely missing . as it is used to separate labels.
+        static $reg_name = '/(?(DEFINE)
+                (?<unreserved> [a-z0-9_~\-])
+                (?<sub_delims> [!$&\'()*+,;=])
+                (?<encoded> %[A-F0-9]{2})
+                (?<reg_name> (?:(?&unreserved)|(?&sub_delims)|(?&encoded)){1,63})
+            )
+            ^(?:(?&reg_name)\.){0,126}(?&reg_name)\.?$/imx';
+
+        static $gen_delims = '/[:\/?#\[\]@ ]/'; // Also includes space.
+
+        if (preg_match($reg_name, $host)) {
+            return true;
         }
 
-        $labels = array_map([$this, 'toAscii'], explode('.', $host));
+        if (preg_match($gen_delims, $host)) {
+            return false;
+        }
 
-        return 127 > count($labels)
-            && 253 > strlen(implode('.', $labels))
-            && $labels === array_filter($labels, [$this, 'isHostLabel']);
+        return (bool) idn_to_ascii($host, IDNA_NONTRANSITIONAL_TO_ASCII, INTL_IDNA_VARIANT_UTS46);
```

Comment:
This should have a note that it's fine due to the above regex. Otherwise this would mark `'0'` as invalid.